Sources/Basic/TerminalController.swift

```diff
@@ -0,0 +1,252 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright 2016 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import libc
+import func POSIX.getenv
+
+private protocol OutputFileStreamProtocol: OutputStream {
+    var fd: Int32 { get }
+    var file: UnsafeMutablePointer<FILE> { get }
+}
+
+final public class OutputFileStream: OutputFileStreamProtocol {
+    private let fd: Int32
+    private let file: UnsafeMutablePointer<FILE>
+
+    init(file: UnsafeMutablePointer<FILE>) {
+        self.file = file
+        self.fd = fileno(file)
+    }
+
+    public func write(_ string: String) {
+        _ = fputs(string, file)
+    }
+}
+
+public var stdout = OutputFileStream(file: libc.stdout)
+public var stderr = OutputFileStream(file: libc.stderr)
+
+/// Terminal color choices.
+public enum TermColor {
+    case noColor
+
+    case red
+    case green
+    case yellow
+
+    case white
+    case black
+    case grey
+
+    /// Returns the color code which can be prefixed on a string to display it in that color.
+    /// Note: To avoid unexpected behavior the color codes are private and not exposed to clients
+    /// other than TerminalController.
+    private var string: String {
+        switch self {
+        case .noColor: return ""
+        case .red: return "\u{001B}[31m"
+        case .green: return "\u{001B}[32m"
+        case .yellow: return "\u{001B}[33m"
+        case .white: return "\u{001B}[37m"
+        case .black: return "\u{001B}[30m"
+        case .grey: return "\u{001B}[30;1m"
+        }
+    }
+}
+
+/// Number of the columns in the current terminal.
+// FIXME: What if user resizes terminal during execution?
+public var terminalWidth: Int = {
+    // Try to get from enviornment.
+    if let columns = POSIX.getenv("COLUMNS"), let width = Int(columns) {
+        return width
+    }
+
+    // Try determining using ioctl.
+    var ws = winsize()
+    if ioctl(1, UInt(TIOCGWINSZ), &ws) == 0 {
+        return Int(ws.ws_col)
+    }
+
+    // If we can't determine, just assume a default.
+    return 80
+}()
+
+/// A class to have better control on output streams especially standard output and standard error streams.
+/// Allows operations like cursor movement and colored text output if allowed by the tty.
+public final class TerminalController<Target: OutputStream> {
+
+    /// A "dumb" terminal indicates that it has no capabailities for eg: update current line, colors
+    /// The output stream is probably a file or similar.
+    public let dumb: Bool
```

Comment:
I wonder if it would be better to have two different protocols for Dumb vs Non-dumb terminal, and make clients get an instance for the current FD. The non-dumb protocol would be  a superset of the other protocol, but higher level clients would need to check (i.e. cast) if they are available.

This would avoid having the `if dumb` repeated in here.

Just an idea, I'm not really sure which is better...